numpy/_core/include/numpy/ndarraytypes.h

```diff
@@ -1667,11 +1722,24 @@ PyArray_CLEARFLAGS(PyArrayObject *arr, int flags)
 #define PyDataType_ISUSERDEF(obj) PyTypeNum_ISUSERDEF(((PyArray_Descr*)(obj))->type_num)
 #define PyDataType_ISEXTENDED(obj) PyTypeNum_ISEXTENDED(((PyArray_Descr*)(obj))->type_num)
 #define PyDataType_ISOBJECT(obj) PyTypeNum_ISOBJECT(((PyArray_Descr*)(obj))->type_num)
-#define PyDataType_HASFIELDS(obj) (((PyArray_Descr *)(obj))->names != NULL)
-#define PyDataType_HASSUBARRAY(dtype) ((dtype)->subarray != NULL)
+#define PyDataType_HASFIELDS(obj) (PyDataType_ISLEGACY((PyArray_Descr*)(obj)) && ((_PyArray_LegacyDescr *)(obj))->names != NULL)
+#define PyDataType_HASSUBARRAY(dtype) (PyDataType_ISLEGACY(dtype) && ((_PyArray_LegacyDescr *)dtype)->subarray != NULL)
 #define PyDataType_ISUNSIZED(dtype) ((dtype)->elsize == 0 && \
                                       !PyDataType_HASFIELDS(dtype))
 #define PyDataType_MAKEUNSIZED(dtype) ((dtype)->elsize = 0)
+/* Access macros for legacy fields */
+// TODO: Public access to these would be preferable only through a function.
+//       (although macros are good enough to rewire in the future).
+#define PyDataType_SUBARRAY(dtype) (!PyDataType_ISLEGACY(dtype) ?  \
+            NULL : ((_PyArray_LegacyDescr *)dtype)->subarray)
+#define PyDataType_NAMES(dtype) (!PyDataType_ISLEGACY(dtype) ?  \
+            NULL : ((_PyArray_LegacyDescr *)dtype)->names)
+#define PyDataType_FIELDS(dtype) (!PyDataType_ISLEGACY(dtype) ?  \
+            NULL : ((_PyArray_LegacyDescr *)dtype)->fields)
+#define PyDataType_METADATA(dtype) (!PyDataType_ISLEGACY(dtype) ?  \
+            NULL : ((_PyArray_LegacyDescr *)dtype)->metadata)
+#define PyDataType_C_METADATA(dtype) (!PyDataType_ISLEGACY(dtype) ?  \
+            NULL : ((_PyArray_LegacyDescr *)dtype)->c_metadata)
```

Comment:
what about a comprimise: make them static functions so they can be properly typed?

Reply:
Sounds good, that also means I enforce the direct access for when it is already cast to `_PyArray_LegacyDescr *`.  (Won't touch the old macros here, as it is more tedious, although I can see us changing them also.)

Reply:
ok